File: 110-balanced-binary-tree/balanced-binary-tree.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
#include <stdlib.h>

int height(struct TreeNode* root) {
    if (!root)
        return -1;
    int left = height(root->left);
    int right = height(root->right);
    return 1 + (left > right ? left : right);
}

bool isBalanced(struct TreeNode* root) {
    if (!root)
        return true;
    if (abs(height(root->left) - height(root->right)) > 1)
        return false;

    return isBalanced(root->left) && isBalanced(root->right);
}

File: 110-balanced-binary-tree/balanced-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def height(self, root: Optional[TreeNode]) -> int:
        if not root:
            return -1
        left = self.height(root.left)
        right = self.height(root.right)
        return 1 + max(left, right)

    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        if not root:
            return True
        if abs(self.height(root.left) - self.height(root.right)) > 1:
            return False
        return self.isBalanced(root.left) and self.isBalanced(root.right)
